cub/cub/detail/nvtx.cuh

```diff
@@ -0,0 +1,76 @@
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#pragma once
+
+#include <cub/config.cuh>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/optional>
+
+#if __has_include(<source_location> )
+#  include <source_location>
+#endif // __has_include(<source_location>)
+
+// NVTX C++ wrapper documentation: https://nvidia.github.io/NVTX/doxygen-cpp/index.html
+#if __has_include(<nvtx3/nvtx3.hpp>) // TODO(bgruber): replace by a check for the first CTK version shipping the header
+#  include <nvtx3/nvtx3.hpp>
+#else // __has_include(<nvtx3/nvtx3.hpp>)
+#  include "nvtx3.hpp"
+#endif // __has_include(<nvtx3/nvtx3.hpp>)
+
+CUB_NAMESPACE_BEGIN
+namespace detail
+{
+struct NVTXCCCLDomain
+{
+  static constexpr const char* name = "CCCL";
+};
+
+// Conditionally inserts a NVTX range starting here until the end of the current function scope in host code. Does
+// nothing in device code.
+// The optional is needed to defer the construction of an NVTX range (host-only code) and message string registration
+// into a dispatch region running only on the host, while preserving the semantic scope where the range is declared.
+#define CUB_NVTX_RANGE_SCOPE_IF(condition, name)                                                                      \
```

Comment:
question: Should we make `CUB_NVTX_RANGE_SCOPE_IF` a no-op if `NVTX_DISABLE` is defined?

Reply:
Since nvtx3 already disables a lot of stuff internally when `NVTX_DISABLE` is defined, I didn't bother first. But in case I missed something, let's add it, since it should be easy.